cryptarchia/cryptarchia.py

```diff
@@ -383,39 +393,65 @@ def __init__(self, genesis_state: LedgerState, config: Config):
 
     def validate_header(self, block: BlockHeader, chain: Chain) -> bool:
         # TODO: verify blocks are not in the 'future'
-        current_state = self.ledger_state[chain.tip_id()].copy()
-        orphaned_commitments = set()
+        if block.parent != chain.tip_id():
+            print("WARN: block parent is not chain tip")
+            return False
+
+        current_state = self.ledger_state[block.parent].copy()
+
         # first, we verify adopted leadership transactions
-        for proof in block.orphaned_proofs:
-            proof = proof.leader_proof
-            # each proof is validated against the last state of the ledger of the chain this block
-            # is being added to before that proof slot
-            parent_state = self.state_at_slot_beginning(chain, proof.slot).copy()
-            # we add effects of previous orphaned proofs to the ledger state
-            parent_state.commitments_lead |= orphaned_commitments
-            epoch_state = self.compute_epoch_state(proof.slot.epoch(self.config), chain)
-            if self.verify_slot_leader(
-                proof.slot, proof, epoch_state, parent_state, current_state
+        for orphan in block.orphaned_proofs:
+            # orphan proofs are checked in two ways
+            # 1. ensure they are valid locally in their original branch
+            # 2. ensure it does not conflict with current state
+
+            # We take a shortcut for (1.) by constraining orphans to proofs we've already processed in other branches
+            if orphan.id() not in self.ledger_state:
+                print("WARN: missing orphan proof")
+                return False
+
+            # we use the proposed block epoch state here instead of the orphan's epoch state.
+            # for very old orphans, these states may be different.
+            epoch_state = self.compute_epoch_state(block.slot.epoch(self.config), chain)
+
+            # each proof is validated against the current tip state, this will change
+            # once we introduce merkle roots in headers.
+            parent_state = current_state
+
+            # (2.) is checked by verifying the proof against the current state ensuring:
+            # - it is a valid proof
+            # - and the nullifier has not already been spent
+            if not self.verify_slot_leader(
+                orphan.slot,
+                orphan.parent,
+                orphan.leader_proof,
+                epoch_state,
+                parent_state,
+                current_state,
             ):
-                # if an adopted leadership proof is valid we need to apply its effects to the ledger state
-                orphaned_commitments.add(proof.evolved_commitment)
-                current_state.nullifiers.add(proof.nullifier)
-            else:
-                print("WARN: orphan proof is invalid")
-                # otherwise, the whole block is invalid
+                print("WARN: invalid orphan proof")
```

Comment:
Is this library code, I thought it was just for use as a reference? I did intend to leave these in, it was helpful in debugging

Reply:
It's not well organized to be used as a library atm but it could be used for that as well. How difficult would it be to switch to some logging library where reporting can be configured?

Reply:
yeah, I can do that, I've replaced the prints with the python logger 👍

(I don't think we should aim to make this code a library, IMO, the goal should just be pedagogical/elucidating)